examples/vae_simple.py

```diff
@@ -0,0 +1,319 @@
+    return l_enc_mu, l_enc_logsigma, l_dec_mu_list, l_dec_logsigma_list, l_output_list, l_output
+
+# ############################## Main program ################################
+
+# helper function for log-likelihood expression
+def log_likelihood(tgt, mu, ls):
+    return T.sum(-(np.float32(0.5 * np.log(2 * np.pi)) + ls)
+            - 0.5 * T.sqr(tgt - mu) / T.exp(2 * ls))
+
+def main(L=2, z_dim=2, n_hid=1024, num_epochs=300, binary=True):
+    print("Loading data...")
+    X_train, X_val, X_test = load_dataset()
+    width, height = X_train.shape[2], X_train.shape[3]
+    input_var = T.tensor4('inputs')
+
+    # Create VAE model
+    print("Building model and compiling functions...")
+    print("L = {}, z_dim = {}, n_hid = {}, binary={}".format(L, z_dim, n_hid, binary))
+    x_dim = width * height
+    l_z_mu, l_z_ls, l_x_mu_list, l_x_ls_list, l_x_list, l_x = \
+           build_vae(input_var, L=L, binary=binary, z_dim=z_dim, n_hid=n_hid)
+
+    # If there are dropout layers etc these functions return masked or non-masked expressions
+    # depending on if they will be used for training or validation/test err calcs
+    layer_outputs = lambda b: nn.layers.get_output([l_z_mu, l_z_ls] +
+            l_x_mu_list + l_x_ls_list + l_x_list, deterministic=b)
+    z_mu = lambda b: layer_outputs(b)[0]
+    z_ls = lambda b: layer_outputs(b)[1]
+    x_mu = lambda b: [] if binary else layer_outputs(b)[2:2+L]
+    x_ls = lambda b: [] if binary else layer_outputs(b)[2+L:2+2*L]
+    x_list = lambda b: layer_outputs(b)[2 if binary else 2+2*L:]
+
+    # Loss expression has two parts as specified in [1]
+    # kl_div = KL divergence between p_theta(z) and p(z|x)
+    # - divergence between prior distr and approx posterior of z given x
+    # - or how likely we are to see this z when accounting for Gaussian prior
+    # logpxz = log p(x|z)
+    # - log-likelihood of x given z
+    # - in binary case logpxz = cross-entropy
+    # - in continuous case, is log-likelihood of seeing the target x under the
+    #   Gaussian distribution parameterized by dec_mu, sigma = exp(dec_logsigma)
+    kl_div = lambda b: 0.5 * T.sum(1 + 2*z_ls(b) - T.sqr(z_mu(b)) - T.exp(2 * z_ls(b)))
+    if binary:
+        logpxz = lambda b: sum(nn.objectives.binary_crossentropy(x,
+            input_var.flatten(2)).sum() for x in x_list(b)) * (-1./L)
+        test_prediction = nn.layers.get_output(l_x, deterministic=True)
+    else:
+        logpxz = lambda b: sum(log_likelihood(input_var.flatten(2), mu, ls)
+            for mu, ls in zip(x_mu(b), x_ls(b)))/L
+        test_prediction = T.sum(x_mu(True), axis=0)/L
+    loss = -1 * (logpxz(False) + kl_div(False))
+    test_loss = -1 * (logpxz(True) + kl_div(True))
+
+    # functions for generating images given a code (used for visualization)
+    # we give the function a certain z and then set logsigma(z) to a very negative
+    # value to make the sampling nearly deterministic
```

Comment:
Is that better than making it fully deterministic? `GaussianSampleLayer` could skip the additive noise for some keyword in `get_output_for()`.